Fix node selection for disruptive tests

abhijat opened the pull request:
## Cover letter

The action injector selects the next node for action using randomness from a set of available nodes. For reversible actions, the node which was affected is added back to available nodes once the action is
reversed.

This leads to a situation where the same node can be selected again and again for disruptions, especially when the node count is low and so randomness is not as effective.

This change switches to a priority queue, initially all nodes begin with priority in the same range so a random node is selected for the first action. once a node is selected, during reversal it is added back with a low priority, so it will not be selected again until all the remaining nodes have been processed.

The priority of nodes is grouped into generations, so that for a 3 node cluster the three nodes initially have the same range of priority, then the next three nodes added back have the same range of priority etc. This ensures that nodes are chosen randomly each time from within nodes of the same generation, and nodes of lower generation are processed first.

Fixes #6889 by making sure that the same node is not restarted repeatedly causing segments not to be removed from it and causing the test to fail.

## Backport Required

<!-- Specify which branches this should be backported to, e.g.: -->
- [ ] not a bug fix
- [ ] issue does not exist in previous branches
- [ ] papercut/not impactful enough to backport
- [x] v22.2.x
- [x] v22.1.x
- [ ] v21.11.x

## UX changes

* none

## Release notes

* none

### Improvements

* fixes failing test.

[force push](https://github.com/redpanda-data/redpanda/compare/8399c2ba7e4ca55ed7203a29bf52e00a05a94ed3..218c537cfa60b802aed10fe0e1b9d956e8bf12a9) - remove time limit changes, fix priority calculation

[force push](https://github.com/redpanda-data/redpanda/compare/218c537cfa60b802aed10fe0e1b9d956e8bf12a9..5ec652456c9affc52e68eea688c148ac5de83ec2) - handle full queue by skipping adding node

[force push](https://github.com/redpanda-data/redpanda/compare/5ec652456c9affc52e68eea688c148ac5de83ec2..0d189a2490a226e08cf001409c016d48e2a3a543) - use simpler node selection, adjust sleep interval between cycles

abhijat commented:
/ci-repeat 2
skip-units
dt-repeat=100
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTestWithDisruptions.test_write_with_node_failures

abhijat commented:
/ci-repeat 2
skip-units
dt-repeat=200
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTestWithDisruptions.test_write_with_node_failures

abhijat commented:
/ci-repeat

abhijat commented:
> /ci-repeat

I ran this assuming cancelled runs like https://buildkite.com/redpanda/redpanda/builds/17509#01841da4-20c5-40e8-8792-c1765002db80 would restart but it looks like only release and debug have started in https://buildkite.com/redpanda/redpanda/builds/17527

ztlpn commented on tests/rptest/services/action_injector.py:
```diff
@@ -42,13 +52,20 @@ class DisruptiveAction:
         self.admin = admin
         self.config = config
         self.redpanda = redpanda
+        self.nodes_for_action = queue.PriorityQueue(
```
nit: PriorityQueue seems like an overkill because we don't need any synchronization here. Although if its interface is more convenient than the raw `heapq` module, then it is okay I guess.

abhijat replied:
Yes, it is mostly for the PriorityQueue API, we always read from it in non-blocking mode although now looking at the code we have max size and we put in blocking mode, maybe it should be fixed. 

If you feel `heapq` should be used let me know, I can try a refactor.

ztlpn commented on tests/rptest/services/action_injector.py:
```diff
@@ -224,10 +243,21 @@ def create_context_with_defaults(redpanda: RedpandaService,
+
+    if op_type.is_reversible:
+        min_time_between_actions = max(int(1 / redpanda.num_nodes * 60), 10)
+        max_time_between_actions = min_time_between_actions * 2
+    else:
+        min_time_between_actions = 10
+        max_time_between_actions = 30
+    redpanda.logger.info(
+        f'Initializing ActionCtx with '
+        f'min time between actions: {min_time_between_actions}, '
+        f'max time between actions: {max_time_between_actions}')
     config = config or ActionConfig(
         cluster_start_lead_time_sec=20,
-        min_time_between_actions_sec=10,
-        max_time_between_actions_sec=30,
+        min_time_between_actions_sec=min_time_between_actions,
+        max_time_between_actions_sec=max_time_between_actions,
```
Looks like this change and the priority queue change are independent. If that is the case, they should be in separate commits.

abhijat replied:
I have removed this change now, initially I thought it would help with the issue but now I do not think adjusting time may be necessary. I can revisit this if required.

ztlpn commented on tests/rptest/services/action_injector.py:
```diff
@@ -42,13 +52,20 @@ class DisruptiveAction:
+        self.nodes_for_action = queue.PriorityQueue(
+            maxsize=self.redpanda.num_nodes)
+        now = int(time.time())
+        for node in self.redpanda.nodes:
+            self.nodes_for_action.put(NodePriority(now, node))
```
Does this mean that there is no randomness in the order in which we choose nodes? Should we at least add them to the queue in the random order?

abhijat replied:
This was an error on my part, I assumed that items with same priority will be returned in random order by the queue, but it is incorrect. Now I have added a utility to get random priority so that nodes are grouped according to the count of disruptive actions performed on them, and nodes with same count have the same range of priority. As the count increases the priority range also increases.

ztlpn reviewed (commented):
Code looks good, although the motivation is a bit unclear to me. Can you explain in more detail how the changes (and which one of them) fix #6889 ?

abhijat commented:
/ci-repeat 2
skip-units
dt-repeat=100
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTestWithDisruptions.test_write_with_node_failures

abhijat commented:
> Code looks good, although the motivation is a bit unclear to me. Can you explain in more detail how the changes (and which one of them) fix #6889 ?

Sure, I have added comments and an explanation to the PR description now. Basically we want to make sure that the node selected for restart is
1. randomly selected
2. once a node is restarted it gets some time after that where it will not be restarted again

In the existing code `1` was fulfilled but after restart, we added the recovered node back to the set of nodes available for action immediately, so there was a chance it will be restarted again soon. 

With the current change a node is put to the back of the queue once restarted, so we do not pick it again until all its peers have been restarted once.

abhijat commented:
/ci-repeat 2
skip-units
dt-repeat=100
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTestWithDisruptions.test_write_with_node_failures

ztlpn commented on tests/rptest/services/action_injector.py:
```diff
@@ -33,6 +34,41 @@ def time_between_actions(self) -> float:
+        if self._items > self._max_items:
+            self._items = self._items % self._max_items
+            self._generation += 1
+        return random.randint(self._generation * self._step,
+                              (self._generation + 1) * self._step)
```
Nit: looks like because for `randint` the boundaries are inclusive, we have a small chance of overlapping priorities for different generations here. So it is better to use `randrange` (and comment needs to be fixed a bit too).

ztlpn commented on tests/rptest/services/action_injector.py:
```diff
@@ -60,19 +126,18 @@ def max_affected_nodes_reached(self) -> bool:
-        Randomly selects the next node to apply the action on. A set of affected
-        nodes is maintained so that we do not apply the action on nodes which were
-        already targeted in previous invocations.
+        Selects the next node to process from a priority queue. Initially all nodes have
+        the same priority. When a node has been processed it gets a much lower priority
+        so that it does not get selected again until all the non-affected nodes have been
+        processed.
```
Not sure if this is the same thing but I thought that maybe we can just keep two lists for the current and the next generation. We pop off current and put into next, and if current becomes empty, we swap next for current and shuffle it. Maybe this is simpler.

abhijat replied:
Yes, this is much simpler, thanks for the suggestion. I tried it out with two sets and it seems to work. Made the change now.

abhijat commented:
Unfortunately there is still a small chance that the same node gets selected at the end of one cycle and then at the beginning of the next (docker-rp-4 near the end):

```
[INFO  - 2022-10-28 17:41:03,163 - action_injector - target_node - lineno:136]: selected docker-rp-4, priority 160
[INFO  - 2022-10-28 17:41:28,800 - action_injector - target_node - lineno:136]: selected docker-rp-8, priority 347
[INFO  - 2022-10-28 17:41:53,962 - action_injector - target_node - lineno:136]: selected docker-rp-7, priority 364
[INFO  - 2022-10-28 17:42:11,103 - action_injector - target_node - lineno:136]: selected docker-rp-8, priority 1089
[INFO  - 2022-10-28 17:42:29,213 - action_injector - target_node - lineno:136]: selected docker-rp-7, priority 1311
[INFO  - 2022-10-28 17:42:56,374 - action_injector - target_node - lineno:136]: selected docker-rp-4, priority 1484
[INFO  - 2022-10-28 17:43:21,561 - action_injector - target_node - lineno:136]: selected docker-rp-4, priority 2331
[INFO  - 2022-10-28 17:43:50,791 - action_injector - target_node - lineno:136]: selected docker-rp-7, priority 2597
```

And this results in a test failure. I have added an increased sleep interval for the iteration when a cycle is ended.

abhijat commented:
/ci-repeat 1
skip-units
dt-repeat=200
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTestWithDisruptions.test_write_with_node_failures

abhijat commented:
/ci-repeat 1
skip-units
dt-repeat=500
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTestWithDisruptions.test_write_with_node_failures

abhijat commented:
/ci-repeat 1
skip-units
dt-repeat=300
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTestWithDisruptions.test_write_with_node_failures

abhijat commented:
/backport v22.2.x

abhijat commented:
/backport v22.1.x
